src/lib/4337.js

```diff
@@ -0,0 +1,249 @@
+
+const accountFactoryInterface = new ethers.utils.Interface(
+  accountFactoryContractAbi
+);
+const accountInterface = new ethers.utils.Interface(accountContractAbi);
+const erc20Token = new ethers.utils.Interface(tokenContractAbi);
+
+const adjustForDecimals = (amount, decimals) => {
+  return amount * 10 ** decimals;
+};
+
+const transferCallData = (tokenAddress, receiver, amount) =>
+  accountInterface.encodeFunctionData("execute", [
+    tokenAddress,
+    ethers.constants.Zero,
+    erc20Token.encodeFunctionData("transfer", [receiver, amount]),
+  ]);
+
+const userOpToJson = (userop) => {
+  const newUserop = {
+    sender: userop.sender,
+    nonce: userop.nonce.toHexString().replace("0x0", "0x"),
+    initCode: ethers.utils.hexlify(userop.initCode),
+    callData: ethers.utils.hexlify(userop.callData),
+    callGasLimit: userop.callGasLimit.toHexString().replace("0x0", "0x"),
+    verificationGasLimit: userop.verificationGasLimit
+      .toHexString()
+      .replace("0x0", "0x"),
+    preVerificationGas: userop.preVerificationGas
+      .toHexString()
+      .replace("0x0", "0x"),
+    maxFeePerGas: userop.maxFeePerGas.toHexString().replace("0x0", "0x"),
+    maxPriorityFeePerGas: userop.maxPriorityFeePerGas
+      .toHexString()
+      .replace("0x0", "0x"),
+    paymasterAndData: ethers.utils.hexlify(userop.paymasterAndData),
+    signature: ethers.utils.hexlify(userop.signature),
+  };
+
+  return newUserop;
+};
+
+const userOpFromJson = (userop) => {
+  const newUserop = {
+    sender: userop.sender,
+    nonce: ethers.BigNumber.from(userop.nonce),
+    initCode: ethers.utils.arrayify(userop.initCode),
+    callData: ethers.utils.arrayify(userop.callData),
+    callGasLimit: ethers.BigNumber.from(userop.callGasLimit),
+    verificationGasLimit: ethers.BigNumber.from(userop.verificationGasLimit),
+    preVerificationGas: ethers.BigNumber.from(userop.preVerificationGas),
+    maxFeePerGas: ethers.BigNumber.from(userop.maxFeePerGas),
+    maxPriorityFeePerGas: ethers.BigNumber.from(userop.maxPriorityFeePerGas),
+    paymasterAndData: ethers.utils.arrayify(userop.paymasterAndData),
+    signature: ethers.utils.arrayify(userop.signature),
+  };
+
+  return newUserop;
+};
+
+const senderAccountExists = async (indexer, sender) => {
+  const url = `${indexer.url}/accounts/${sender}/exists`;
+
+  const resp = await fetch(url);
+
+  return resp.status === 200;
+};
+
+const generateUserOp = (
+  signerAddress,
+  sender,
+  senderAccountExists = false,
+  accountFactoryAddress,
+  callData
+) => {
+  const userop = {
+    sender,
+    nonce: ethers.BigNumber.from("0"),
+    initCode: ethers.utils.arrayify("0x"),
+    callData: ethers.utils.arrayify("0x"),
+    callGasLimit: ethers.BigNumber.from("0"),
+    verificationGasLimit: ethers.BigNumber.from("1500000"),
+    preVerificationGas: ethers.BigNumber.from("21000"),
+    maxFeePerGas: ethers.BigNumber.from("0"),
+    maxPriorityFeePerGas: ethers.BigNumber.from("1000000000"),
```

Comment:
Does that mean that those values are hard coded in the userop and cannot changed by the paymaster?

I thought it was the responsibility of the paymaster to pay the right amount to make sure the userop is recorded on chain.

Reply:
These values are used for user op simulation. A few of them are actually updated and returned by the paymaster RPC. This is all so that an accurate refund can be calculated into the deposit on the entrypoint on chain.

Since we made our own simplified entrypoint, this is not needed. Our userop RPC makes sure to calculate the appropriate amount of gas when submitting userops.

Reply:
Just respecting the interface.